core/server/master/src/main/java/alluxio/master/file/meta/MountTable.java

```diff
@@ -104,47 +117,98 @@ public MountTable(UfsManager ufsManager, MountInfo rootMountInfo) {
+   * NOT ThreadSafe. This method will not acquire any locks, so the caller MUST apply the lock
+   * first before calling this method.
+   * @param alluxioUri the alluxio path that is about to be the mountpoint
+   * @param ufsUri the UFS path that is about to mount
+   */
+  public void validateMountPoint(AlluxioURI alluxioUri, AlluxioURI ufsUri)
+      throws FileAlreadyExistsException, InvalidPathException, IOException {
+    String alluxioPath = alluxioUri.getPath().isEmpty() ? "/" : alluxioUri.getPath();
+    LOG.info("Validating Mounting {} at {}", ufsUri, alluxioPath);
+    if (mState.getMountTable().containsKey(alluxioPath)) {
+      throw new FileAlreadyExistsException(
+          ExceptionMessage.MOUNT_POINT_ALREADY_EXISTS.getMessage(alluxioPath));
+    }
+    // Make sure that the ufs path we're trying to mount is not a prefix
+    // or suffix of any existing mount path.
+    for (Map.Entry<String, MountInfo> entry : mState.getMountTable().entrySet()) {
+      AlluxioURI mountedUfsUri = entry.getValue().getUfsUri();
+      if ((ufsUri.getScheme() == null || ufsUri.getScheme().equals(mountedUfsUri.getScheme()))
+          && (ufsUri.getAuthority().toString().equals(mountedUfsUri.getAuthority().toString()))) {
+        String ufsPath = ufsUri.getPath().isEmpty() ? "/" : ufsUri.getPath();
+        String mountedUfsPath = mountedUfsUri.getPath().isEmpty() ? "/" : mountedUfsUri.getPath();
+        if (PathUtils.hasPrefix(ufsPath, mountedUfsPath)) {
+          throw new InvalidPathException(ExceptionMessage.MOUNT_POINT_PREFIX_OF_ANOTHER
+              .getMessage(mountedUfsUri.toString(), ufsUri.toString()));
+        }
+        if (PathUtils.hasPrefix(mountedUfsPath, ufsPath)) {
+          throw new InvalidPathException(ExceptionMessage.MOUNT_POINT_PREFIX_OF_ANOTHER
+              .getMessage(ufsUri.toString(), mountedUfsUri.toString()));
         }
       }
+    }
 
-      Map<String, String> properties = options.getPropertiesMap();
-      mState.applyAndJournal(journalContext, AddMountPointEntry.newBuilder()
-          .addAllProperties(properties.entrySet().stream()
-              .map(entry -> StringPairEntry.newBuilder()
-                  .setKey(entry.getKey()).setValue(entry.getValue()).build())
-              .collect(Collectors.toList()))
-          .setAlluxioPath(alluxioPath)
-          .setMountId(mountId)
-          .setReadOnly(options.getReadOnly())
-          .setShared(options.getShared())
-          .setUfsPath(ufsUri.toString())
-          .build());
+    // Check that the alluxioPath we're creating doesn't shadow a path in the parent UFS
+    MountTable.Resolution resolution = resolve(alluxioUri);
+    try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
+      String ufsResolvedPath = resolution.getUri().getPath();
+      if (ufsResource.get().exists(ufsResolvedPath)) {
+        throw new IOException(MessageFormat.format(
```

Comment:
Why `IOException`? Is `InvalidPathException` a better choice?

Reply:
^